editor/src/messages/portfolio/document/node_graph/node_graph_message_handler.rs

```diff
@@ -2586,34 +2547,31 @@ fn frontend_inputs_lookup(breadcrumb_network_path: &[NodeId], network_interface:
 		return Default::default();
 	};
 	let mut frontend_inputs_lookup = HashMap::new();
-	for (&node_id, node) in network.nodes.iter() {
-		let mut inputs = Vec::with_capacity(node.inputs.len());
-		for (index, input) in node.inputs.iter().enumerate() {
-			let is_exposed = input.is_exposed_to_frontend(breadcrumb_network_path.is_empty());
-
-			// Skip not exposed inputs (they still get an entry to help with finding the primary input)
-			if !is_exposed {
-				inputs.push(None);
-				continue;
-			}
-
+	for (node_id, index, output_connector, is_exposed) in network
+		.nodes
+		.iter()
+		.flat_map(|(node_id, node)| {
+			node.inputs
+				.iter()
+				.enumerate()
+				.map(|(index, input)| (*node_id, index, OutputConnector::from_input(input), input.is_exposed()))
+		})
+		.collect::<Vec<_>>()
+	{
+		// Skip not exposed inputs (they still get an entry to help with finding the primary input)
+		let lookup = if !is_exposed {
+			None
+		} else {
```

Comment:
refactor collecting frontend metadata for node inputs when sending the node graph to the frontend.